### PR TITLE
[+] Command Generate - Generate MongoDB nested objects and subDocuments.

### PR DESCRIPTION
**Pull request by SteveBunlon**

### CODE REVIEW

**Comment by SteveBunlon** [comment]
[Lumber - Detect deep fields in embedded documents and array of embedded documents.](https://trello.com/c/Rl23LbZE/3900-lumber-detect-deep-fields-in-embedded-documents-and-array-of-embedded-documents)

File: services/analyzer/mongo-embedded-analyzer.js [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+
+  return parsedFieldAnalyses;
+}
+
+function mergeEmbeddedDetections(keyAnalyses) {
+  if (!areAnalysesSameEmbeddedType(keyAnalyses)) {
+    return 'Object';
+  }
+  const firstAnalysis = keyAnalyses[0];
+  var structure;
```
Comment (slimee) [inline services/analyzer/mongo-embedded-analyzer.js]: We may use let instead of var I think

File: test/expected/mongo/db-analysis-output/nested-array-of-numbers-fields.json [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        "type": "String"
+      },
+      {
+        "name": "propArrayOfNumbers",
+        "type": [
+          "Number",
+          "Number",
+          "Number"
+        ]
+      }
```
Comment (slimee) [inline test/expected/mongo/db-analysis-output/nested-array-of-numbers-fields.json]: We should have "Number" only one time.

File: test/expected/mongo/db-analysis-output/nested-array-of-objects-fields.json [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+            "_id": "mongoose.Schema.Types.ObjectId",
+            "one": "String",
+            "two": "String"
+          },
+          {
+            "_id": "mongoose.Schema.Types.ObjectId",
+            "one": "String",
+            "two": "String"
+          }
+        ]
```
Comment (slimee) [inline test/expected/mongo/db-analysis-output/nested-array-of-objects-fields.json]: We should have one object in the array: `[{_id,one,two}]`

File: test/expected/mongo/db-analysis-output/deep-nested-fields.json [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        "name": "very",
+        "type": {
+          "deep": {
+            "model": {
+              "arrayMixed": [
+                "Number",
+                "String",
+                "Boolean",
+                "Date"
+              ],
```
Comment (slimee) [inline test/expected/mongo/db-analysis-output/deep-nested-fields.json]: Should we better have "arrayMixed": ["Mixed"] here?

File: test/expected/mongo/db-analysis-output/deep-nested-fields.json [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    "fields": [
+      {
+        "name": "name",
+        "type": "String"
+      },
+      {
+        "name": "very",
+        "type": {
+          "deep": {
+            "model": {
```
Comment (slimee) [inline test/expected/mongo/db-analysis-output/deep-nested-fields.json]: `deep` and `model` prop do not have the same model than their parent, `very`
`very' is the value of 'name' prop, but `deep` and `model` are directly used as keys.
Is this first level vs deep levels?
At first sight, I think we should have the same structure independently from the level.

File: test/expected/mongo/dumper-output/deep-nested-js [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+// Learn how here: https://docs.forestadmin.com/documentation/v/v5/reference-guide/models/enrich-your-models
+const mongoose = require('mongoose');
+
+// This section contains the properties of your model, mapped to your collection's properties.
+// Learn more here: https://docs.forestadmin.com/documentation/v/v5/reference-guide/models/enrich-your-models#declaring-a-new-field-in-a-model
+const schema = mongoose.Schema({
+  'name': String,
+  'very': mongoose.Schema({
+    deep: mongoose.Schema({
+      model: mongoose.Schema({
```
Comment (slimee) [inline test/expected/mongo/dumper-output/deep-nested-js]: [MISC] I think we don't need `mongoose.Schema`. 
It's needed when the schema is used in many places.

File: test/expected/mongo/dumper-output/nested-array-of-objects-js [hunk 7]
```diff
@@ -0,0 +1,18 @@
+// This model was generated by Lumber. However, you remain in control of your models.
+// Learn how here: https://docs.forestadmin.com/documentation/v/v5/reference-guide/models/enrich-your-models
+const mongoose = require('mongoose');
+
+// This section contains the properties of your model, mapped to your collection's properties.
+// Learn more here: https://docs.forestadmin.com/documentation/v/v5/reference-guide/models/enrich-your-models#declaring-a-new-field-in-a-model
+const schema = mongoose.Schema({
+  'name': String,
+  'propArrayOfObjects': [mongoose.Schema({
```
Comment (slimee) [inline test/expected/mongo/dumper-output/nested-array-of-objects-js]: [MISC] Here too, it seems we don't need `mongoose.Schema`

File: handlerbars/partials/render-nested.js [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+{{~else if (isObject type)}}
+mongoose.Schema({
+{{#each type}}
+{{indent (sum ../level 1) @key}}: {{>renderNested type=this level=(sum ../level 1)}},
+{{/each}}
+{{indent level '})'}}
+{{else}}
+{{type}}
+{{~/if}}
+`,
```
[the remark this answers is not in the document]
Comment (SteveBunlon) [reply handlerbars/partials/render-nested.js]: Not sure about that. I need to write the model recursively (for nested of nested for example). As we already use handlebars to generate the models from templates, the best way to do that is to use partials, and this is the way to declare it 😞

File: services/analyzer/mongo-collections-analyzer.js [hunk 9]
```diff
@@ -13,6 +32,23 @@ function isUnderscored(fields) {
...
+    analyse,
+    getPrimitiveType,
+    isTypePrimitive,
+    hasEmbeddedTypes,
+    mergeEmbeddedDetections,
+    applyType,
+    areAnalysesSameEmbeddedType,
+    haveSameEmbeddedType,
+    deserializeAnalysis,
+    deserializeAnalyses,
```
[the remark this answers is not in the document]
Comment (SteveBunlon) [reply services/analyzer/mongo-collections-analyzer.js]: These will be removed 👍

File: services/analyzer/mongo-embedded-analyzer.js [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+      Object.keys(type).forEach((key) => {
+        applyType(type[key], nestedStructure, key);
+      });
+    }
+  }
+}
+
+function haveSameEmbeddedType(type1, type2) {
+  return typeof type1 === typeof type2
+    && Array.isArray(type1) === Array.isArray(type2);
```
[the remark this answers is not in the document]
Comment (SteveBunlon) [reply services/analyzer/mongo-embedded-analyzer.js]: It does, I added a test for that

File: handlerbars/helpers/is-array.js [hunk 11]
```diff
@@ -0,0 +1,5 @@
+const Handlebars = require('handlebars');
+
+Handlebars.registerHelper('isArray', (value) => {
+  return Array.isArray(value);
+});
```
[the remark this answers is not in the document]
Comment (SteveBunlon) [reply handlerbars/helpers/is-array.js]: Why should I do this ? Is it performance related topic ?

File: handlerbars/helpers/is-object.js [hunk 12]
```diff
@@ -0,0 +1,5 @@
+const Handlebars = require('handlebars');
+
+Handlebars.registerHelper('isObject', (value) => {
+  return typeof value === 'object';
+});
```
[the remark this answers is not in the document]
Comment (SteveBunlon) [reply handlerbars/helpers/is-object.js]: Why should I do this ? Is it performance related topic ?

File: handlerbars/helpers/sum.js [hunk 13]
```diff
@@ -0,0 +1,5 @@
+const Handlebars = require('handlebars');
+
+Handlebars.registerHelper('sum', (value1, value2) => {
+  return value1 + value2;
+});
```
[the remark this answers is not in the document]
Comment (SteveBunlon) [reply handlerbars/helpers/sum.js]: Why should I do this ? Is it performance related topic ?

File: services/analyzer/mongo-collections-analyzer.js [hunk 14]
```diff
@@ -13,6 +32,23 @@ function isUnderscored(fields) {
...
+    serializeAnalysis,
```
[the remark this answers is not in the document]
Comment (SteveBunlon) [reply services/analyzer/mongo-collections-analyzer.js]: Agreed 👍

File: utils/mongo-primitive-type.js [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+function isTypePrimitive(value) {
+  if (typeof value === 'object' && value instanceof Date) {
+    return true;
+  }
+
+  if (typeof value === 'object' && value instanceof ObjectId) {
+    return true;
+  }
+
+  return PRIMITIVE_TYPES.indexOf(typeof value) > -1;
```
[the remark this answers is not in the document]
Comment (SteveBunlon) [reply utils/mongo-primitive-type.js]: 👍

File: utils/mongo-primitive-type.js [hunk 16]
```diff
@@ -0,0 +1,42 @@
+const { ObjectId } = require('mongodb');
+
+const PRIMITIVE_TYPES = ['string', 'number', 'boolean'];
+
+function getPrimitiveType(value) {
```
[the remark this answers is not in the document]
Comment (SteveBunlon) [reply utils/mongo-primitive-type.js]: I'm not sure about that, it tries to retrieve a 'considered as' primitive type. If we can not find a primitive type ('Date', 'String', 'ObjectId', 'Boolean', 'Number') then we return null. I changed the function name to `getMongoseTypeFromValue()`, it is more explicit IMO. Tell me what.

File: test/services/analyse-mongo-embedded.js [hunk 17]
```diff
@@ -0,0 +1,334 @@
...
+
+        analysis = analysePrimitive([]);
+        expect(analysis).to.equal(null);
+
+        analysis = analysePrimitive(undefined);
+        expect(analysis).to.equal(null);
+
+        analysis = analysePrimitive(null);
+        expect(analysis).to.equal(null);
+      });
```
[the remark this answers is not in the document]
Comment (SteveBunlon) [reply test/services/analyse-mongo-embedded.js]: Yes 👍

File: services/analyzer/mongo-embedded-analyzer.js [hunk 18]
```diff
@@ -0,0 +1,231 @@
...
+      analyses.push(analysis);
+    }
+  });
+
+  return analyses.length ? analyses : null;
+}
+
+function analysePrimitive(value) {
+  return getPrimitiveType(value);
+}
```
[the remark this answers is not in the document]
Comment (SteveBunlon) [reply services/analyzer/mongo-embedded-analyzer.js]: Agreed 👍

File: services/analyzer/mongo-embedded-analyzer.js [hunk 19]
```diff
@@ -0,0 +1,231 @@
...
+function serializeAnalysis(fieldAnalysis) {
+  if (!fieldAnalysis) {
+    return null;
+  }
+
+  const analysis = { type: 'embedded' };
+  analysis.schema = JSON.stringify(fieldAnalysis);
+
+  return analysis;
+}
```
[the remark this answers is not in the document]
Comment (SteveBunlon) [reply services/analyzer/mongo-embedded-analyzer.js]: These JSON.stringify / parse utilities have been deleted 👍

File: services/analyzer/mongo-embedded-analyzer.js [hunk 20]
```diff
@@ -0,0 +1,231 @@
...
+
+  return JSON.parse(analysis.schema);
+}
+
+function deserializeAnalyses(fieldAnalyses) {
+  if (!fieldAnalyses || fieldAnalyses.length === 0) {
+    return null;
+  }
+
+  const parsedFieldAnalyses = [...fieldAnalyses];
```
[the remark this answers is not in the document]
Comment (SteveBunlon) [reply services/analyzer/mongo-embedded-analyzer.js]: The deserialize and serialize functions have been removed

File: services/analyzer/mongo-embedded-analyzer.js (hunk 1 above)
Comment (SteveBunlon) [reply services/analyzer/mongo-embedded-analyzer.js]: For some reasons, let doesn't work in map reduce mongo side 😞

File: test/services/analyse-mongo-embedded.js [hunk 21]
```diff
@@ -0,0 +1,334 @@
...
+
+      it('Should detect that the value to analyse is a primitive value', () => {
+        let analysis = analyse('1');
+        expect(analysis).to.equal('String');
+
+        analysis = analyse(1);
+        expect(analysis).to.equal('Number');
+
+        analysis = analyse(true);
+        expect(analysis).to.equal('Boolean');
```
[the remark this answers is not in the document]
Comment (SteveBunlon) [reply test/services/analyse-mongo-embedded.js]: Not possible to use sinon using functions exported from a module :( I added edge cases tests.

File: services/analyzer/mongo-embedded-analyzer.js [hunk 22]
```diff
@@ -0,0 +1,231 @@
...
+  }
+
+  return null;
+}
+
+function hasEmbeddedTypes(analyses) {
+  if (!analyses || !analyses.length) {
+    return false;
+  }
+  return analyses.filter((analysis) => analysis.type === 'embedded').length > 0;
```
[the remark this answers is not in the document]
Comment (SteveBunlon) [reply services/analyzer/mongo-embedded-analyzer.js]: Not sure about that one, I want to return false if analyses are empty, not the contrary.

File: services/analyzer/mongo-embedded-analyzer.js [hunk 23]
```diff
@@ -0,0 +1,231 @@
...
+      structure[currentKey] = nestedStructure;
+      Object.keys(type).forEach((key) => {
+        applyType(type[key], nestedStructure, key);
+      });
+    }
+  }
+}
+
+function haveSameEmbeddedType(type1, type2) {
+  return typeof type1 === typeof type2
```
[the remark this answers is not in the document]
Comment (SteveBunlon) [reply services/analyzer/mongo-embedded-analyzer.js]: In this function, we want to check that the analyses have the same type. So typeof is good there.  I used `Array.isArray(type1) === Array.isArray(type2)` cause `typeof [] === typeof {}` ==> true, which is not correct in our case (typeof Array returns object 😞 )

File: services/analyzer/mongo-collections-analyzer.js [hunk 24]
```diff
@@ -73,7 +119,12 @@ const mapReduceErrors = (resolve, reject, collectionName) => (err, results) => {
     return reject(err);
   }
   /* eslint no-underscore-dangle: off */
-  return resolve(results.map((r) => ({ name: r._id, type: r.value })));
+  return resolve(results.map((r) => {
+    if (r.value && r.value.type === 'embedded') {
+      return { name: r._id, type: deserializeAnalysis(r.value) };
+    }
+    return { name: r._id, type: r.value };
+  }));
```
[the remark this answers is not in the document]
Comment (SteveBunlon) [reply services/analyzer/mongo-collections-analyzer.js]: Done 👍

File: test/expected/mongo/db-analysis-output/deep-nested-fields.json (hunk 5 above)
Comment (SteveBunlon) [reply test/expected/mongo/db-analysis-output/deep-nested-fields.json]: I don't get you there sorry 😥 Ping me to discuss this together

**Comment by slimee** [comment]
It seems the `mongo-embedded-analyzer.js` file is tested by two separated files:
`test/services/analyzer/mongo-embedded-analyzer.js`
and `test/services/analyse-mongo-embedded.js`

**Comment by slimee** [comment]
20 tests are failing on my laptop.
It's about functions in test but not existing in tested code.

Screenshot 2019-12-27 at 12 50 26

**Comment by arnaudbesnier** [comment]
> 20 tests are failing on my laptop.
> It's about functions in test but not existing in tested code.
> 
> Screenshot 2019-12-27 at 12 50 26

I rebased on `devel` and adapted the tests with Jest.
I have no failing test on my end.

File: test/expected/mongo/db-analysis-output/multiple-nested-array-of-objects-fields.json [hunk 25]
```diff
@@ -0,0 +1,26 @@
...
+      {
+        "name": "name",
+        "type": "String"
+      },
+      {
+        "name": "propArrayOfObjects",
+        "type": [
+          {
+            "one": "String",
+            "two": "String"
```
[the remark this answers is not in the document]
Comment (SteveBunlon) [reply test/expected/mongo/db-analysis-output/multiple-nested-array-of-objects-fields.json]: This is now supported by last commits 😃

File: test/expected/mongo/db-analysis-output/multiple-nested-array-of-objects-fields.json [hunk 26]
```diff
@@ -0,0 +1,26 @@
...
+    ],
+    "options": {
+      "timestamps": false
+    },
+    "primaryKeys": [
+      "_id"
+    ],
+    "references": []
+  }
+}
```
[the remark this answers is not in the document]
Comment (SteveBunlon) [reply test/expected/mongo/db-analysis-output/multiple-nested-array-of-objects-fields.json]: Yes it is intended. multiple stands for multiple records. But multiple records / only one record can lead to the same schema generated.

File: test/services/database-analyzer/database-analyzer-mongo.js [hunk 27]
```diff
@@ -8,11 +8,21 @@ const hasManyModel = require('../../fixtures/mongo/hasmany-model');
 const multipleReferencesModel = require('../../fixtures/mongo/multiple-references-same-field-model');
 const manyNullsModel = require('../../fixtures/mongo/many-nulls-model');
 const complexModel = require('../../fixtures/mongo/many-objectid-fields-model');
+const nestedObjectModel = require('../../fixtures/mongo/nested-object-model');
+const nestedArrayOfObjectsModel = require('../../fixtures/mongo/nested-array-of-objects-model');
+const nestedArrayOfNumbersModel = require('../../fixtures/mongo/nested-array-of-numbers-model');
+const deepNestedModel = require('../../fixtures/mongo/deep-nested-model');
+const multipleNestedArrayOfObjectsModel = require('../../fixtures/mongo/nested-array-of-objects-model');
```
[the remark this answers is not in the document]
Comment (SteveBunlon) [reply test/services/database-analyzer/database-analyzer-mongo.js]: Nice catch 💪

File: services/analyzer/mongo-collections-analyzer.js [hunk 28]
```diff
@@ -4,6 +4,18 @@ const logger = require('../logger');
...
 const { detectReferences, applyReferences } = require('./mongo-references-analyzer');
 const { detectHasMany, applyHasMany } = require('./mongo-hasmany-analyzer');
+const {
+  getMongooseTypeFromValue,
+  isOfMongooseType,
+} = require('../../utils/mongo-primitive-type');
+const {
+  getMongooseArraySchema,
+  getMongooseEmbeddedSchema,
+  getMongooseSchema,
```
Comment (slimee) [inline services/analyzer/mongo-collections-analyzer.js]: [IGNORABLE] we are in mongo prefixed files so no need to let mongo in function.

File: services/analyzer/mongo-collections-analyzer.js [hunk 29]
```diff
@@ -47,14 +68,31 @@ function mapCollection() {
...
+        if (analysis) {
+          // Notice: Wrap the analysis of embedded in a recognizable object for further treatment
+          emit(key, { type: 'embedded', schema: analysis });
+        }
       }
     }
   }
 }
 /* eslint-enable */
+function reduceCollection(key, analyses) {
```
Comment (slimee) [inline services/analyzer/mongo-collections-analyzer.js]: [IGNORABLE] thanks you for removing stuff :D

File: services/analyzer/mongo-embedded-analyzer.js [hunk 30]
```diff
@@ -0,0 +1,219 @@
...
+  }
+
+  return null;
+}
+
+function hasEmbeddedTypes(analyses) {
+  if (!analyses || !analyses.length) {
+    return false;
+  }
+  return analyses.filter((analysis) => analysis.type === 'embedded').length > 0;
```
Comment (slimee) [inline services/analyzer/mongo-embedded-analyzer.js]: [IGNORABLE] a one liner here looks not KISS

File: test/services/database-analyzer/database-analyzer-mongo.test.js [hunk 31]
```diff
@@ -89,5 +109,125 @@ describe('services > database analyser > MongoDB', () => {
...
+    it('should generate the model with a nested object', async () => {
+      expect.assertions(1);
+      const mongoHelper = await getMongoHelper(mongoUrl);
+      const databaseConnection = await mongoHelper.connect();
+      await mongoHelper.dropAllCollections();
+      await mongoHelper.given(nestedObjectModel);
+      const databaseAnalyzer = new DatabaseAnalyzer(databaseConnection, { dbDialect: 'mongodb' });
+      const model = await databaseAnalyzer.perform();
+      expect(model).toStrictEqual(expectedNestedObjectModel);
+      await mongoHelper.close();
```
Comment (slimee) [inline test/services/database-analyzer/database-analyzer-mongo.test.js]: [IGNORABLE] This code looks close to others, we may have a function that encapsulate all. 
Something like `databaseAnalyzerTest(databaseConnection, fixtureModel, expectedModel)`

File: test/services/dumper/dumper-mongo.test.js [hunk 32]
```diff
@@ -52,4 +61,103 @@ describe('services > dumper > MongoDB', () => {
...
+
+  it('generate a model file with a nested object', async () => {
+    expect.assertions(1);
+    const dumper = await getDumper();
+    await dumper.dump(nestedObjectModel);
+    const generatedFile = fs.readFileSync('./test-output/mongo/models/persons.js', 'utf8');
+    const expectedFile = fs.readFileSync('./test-expected/mongo/dumper-output/nested-object.expected.js', 'utf-8');
+
+    expect(generatedFile).toStrictEqual(expectedFile);
+    cleanOutput();
```
Comment (slimee) [inline test/services/dumper/dumper-mongo.test.js]: [IGNORABLE] We can encapsulate the common code in a function.

File: services/analyzer/mongo-collections-analyzer.js (hunk 28 above)
Comment (SteveBunlon) [reply services/analyzer/mongo-collections-analyzer.js]: I would rather keep this naming, because what we want is mongoose specific data, not even mongo data. IMO the mongoose word has its own meaning
